### ISSUE TITLE
net: add mechanism to wait for readability on a TCPConn

### ISSUE THREAD

**Issue by bradfitz** [golang/go#15735]
The net/http package needs a way to wait for readability on a TCPConn without actually reading from it. (See #15224)

http://golang.org/cl/22031 added such a mechanism, making Read(0 bytes) do a wait for readability, followed by returning (0, nil). But maybe that is strange. Windows already works like that, though. (See new tests in that CL)

Reconsider this for Go 1.8.

Maybe we could add a new method to TCPConn instead, like WaitRead.

**Comment by RalphCorderoy** [comment]
read(2) with a count of zero may be used to detect errors.  Linux man page confirms, as does POSIX's read(3p) here.  Mentioning it in case it influences this subverting of a Read(0 bytes) not calling syscall.Read.

**Comment by bradfitz** [comment]
I found a way to do without this in net/http, so punting to Go 1.9.

**Comment by bradfitz** [comment]
Actually, the more I think about this, I don't even want my idle HTTP/RPC goroutines to stick around blocked in a read call. In addition to the array memory backed by the slice given to `Read`, the goroutine itself is ~4KB of wasted memory. 

What I'd really like is a way to register a `func()` to run when my `*net.TCPConn` is readable (when a `Read` call wouldn't block). By analogy, I want the `time.AfterFunc` efficiency of running a func in a goroutine later, rather than running a goroutine just to block in a `time.Sleep`.

My new proposal is more like:

```go
package net

// OnReadable runs f in a new goroutine when c is readable;
// that is, when a call to c.Read will not block.
func (c *TCPConn) OnReadable(f func()) {
   // ...
}
```

Yes, maybe this is getting dangerously into event-based programming land.

Or maybe just the name ("`OnWhatever`") is offensive. Maybe there's something better.

I would use this in http, http2, and grpc.

/cc @ianlancetaylor @rsc

**Comment by ianlancetaylor** [comment]
Sounds like you are getting close to #15021.

I'm worried that the existence of such a method will encourage people to start writing their code as callbacks rather than as straightforward goroutines.

**Comment by bradfitz** [comment]
Yeah. I'm conflicted. I see the benefits and the opportunity for overuse.

**Comment by dvyukov** [comment]
If we do OnReadable(f func()), won't we need to fork half of standard library for async style? Compress, io, tls, etc readers all assume blocking style and require a blocked goroutine.
I don't see any way to push something asynchronously into e.g. gzip.Reader. Does this mean that I have to choose between no blocked goroutine + my own gzip impl and blocked goroutine + std lib?

**Comment by dvyukov** [comment]
Re 0-sized reads.
It should work with level-triggered notifications, but netpoll uses epoll in edge-triggered mode (and kqueue iirc). I am concerned if cl/22031 works in more complex cases: waiting for already ready IO, double wait, wait without completely draining read buffer first, etc?

**Comment by bradfitz** [comment]
@dvyukov, no, we would only use `OnReadable` in very high-level places, like the http1 and http2 servers where we know the conn is expected to be idle for long periods of time. The rest of the code underneath would remain in the blocking style.

**Comment by dvyukov** [comment]
This looks like a half-measure. An http connection can halt in the middle of request...

**Comment by bradfitz** [comment]
@dvyukov, but not commonly. This would be an optimization for the common case.

**Comment by dvyukov** [comment]
An alternative interface can be to register a channel that will receive readiness notifications. The other camp wants this for packet-processing servers, and there starting a goroutine for every packet will be too expensive. However, if at the end you want a goroutine, then the channel will introduce unnecessary overhead.
Channel has a problem with overflow handling (netpoll can't block on send, on the other hand it is not OK to lose notifications).
For completeness, this API should also handle writes.

**Comment by DemiMarie** [comment]
We need to make sure that this works with Windows IOCP as well.

**Comment by rsc** [comment]
Not obvious to me why the API has to handle writes. The thing about reads is that until the data is ready for reading, you can use the memory for other work. If you're waiting to write data, that memory is not reusable (otherwise you'd lose the data you are waiting to write).

**Comment by dvyukov** [comment]
@rsc If we do just 0-sized reads, then write support is not necessary. However, if we do Brad's "My new proposal is more like": `func (c *TCPConn) OnReadable(f func())`, then this equally applies to writes as well -- to avoid 2 blocked goroutines per connection.

**Comment by noblehng** [comment]
If memory usage is the concern, it is possible to make long parked G use less memory instead of changing programming style? One main selling point of Go to me is high efficiency network servers without resorting to callbacks.

Something like shrink the stack or move the stack to heap by the GC using some heuristics, that will be littile different from spinning up a new goroutine on callback memory usage wise, and scheduling wise a callback is not much different than `goready()`. Also I assume the liveness change in Go1.8 could help here too.

For the backing array, if it is preallocated buffer, than a callback doesn't make much different than `Read()`, maybe it will make some different if it is allocated per-callback and use a pool.

**Comment by noblehng** [comment]
How about a epoll like interface for `net.Listener`:
```go
type PollableListener interface {
   net.Listener
   // Poll will block till at least one connection been ready for read or write
   // reads and writes are special net.Conn that will not block on EAGAIN
   Poll() (reads []net.Conn, writes []net.Conn)
}
```
Then the caller of `Poll()` can has a small number of goroutines to poll for readiness and handle the reads and writes. This should also works well for packet-processing servers.

Note that this only needs to be implemented in the runtime for those Listeners that multiplexed in the kernel, like the `net.TCPListener`. For other protocol that multiplex in the userspace and doesn't attached to the runtime poller directly, like udp listener or multiplexing streams in a tcp connection, can be implemented outside the runtime. For example, for multiplexing in a tcp connection, we can implemented the epoll like behavior by read from/write to some buffers then poll from them.

**Comment by aajtodd** [comment]
I'm fairly new to Go but seeing the callback interface is a little grating given the blocking API exposed everywhere else. Why not expose a public API to the netpoll interfaces?

Go provides no standard public facing event loop (correct me if I'm wrong please). I have need to wait for readability on external FFI socket(s) (given through cgo). It would be nice to re-use the existing netpoll abstraction to also spawn FFI sockets onto rather than having to wrap epoll/IOCP/select. Also I'm guessing wrapping (e.g) epoll from the sys package does not integrate with the scheduler which would also be a bummer.

**Comment by mjgarton** [comment]
For a number of my use cases, something like this :

```
package net

// Readable returns a channel which can be read from whenever a call to c.Read
// would not block.
func (c *TCPConn) Readable() <-chan struct{} {
        // ...
}

```

.. would be nice because I can select on it. I have no idea whether it's practical to implement this though.

Another alternative (for some of my cases at least) might be somehow enabling reads to be canceled by using a context.

**Comment by funny-falcon** [comment]
```go
type ReadableSig struct {
    C *TCPConn,
    I interface{}
}
// Readable marks connection as waiting for read.
// It will send connection to channel when it becomes readable (or closed by peer)
//  together with additional information passed as 'i' argument.
// Note: if channel blocks, it will create goroutine.
func (c *TCPConn) Readable(i interface{}, ch chan<- ReadableSig) {
}
```

**Comment by funny-falcon** [comment]
Or even better:
```go
type ReadableNotify struct {
    Conn *TCPConn,
    Ctx context.Context
}
// Readable marks connection as waiting for read and It will send connection
// to channel when it becomes readable (or closed) together with context.
// Internally it may spawn goroutine. It will certainly spawn goroutine if channel
// blocks.
// Note: connections always waits for readability, but if context is closed then
// notification may be omitted. If notification recieved, then Context should be
// checked manually.
func (c *TCPConn) Readable(ctx context.Context, ch chan<- ReadableNotify)
```
btw, can runtime spawn utility goroutine with less stack than user-faced goroutine?

**Comment by gobwas** [comment]
Hi! Great idea 🎆

In our go server with >2millions of alive WebSocket connections we were used epoll to bring the same functionality.

```go
epoll.Handle(conn, EPOLLIN | EPOLLONESHOT, func(events uint) {
    // No more calls will be made for conn until we call epoll.Resume().
    ...
    epoll.Resume(conn)
})
```

That is, it could be useful to make callback call only once after registration.
In our case we've used EPOLLONESHOT to make this work.

**Comment by amits1995** [comment]
I think you'll have to add support for Registered IO in windows, instead of IO Completion Ports

**Comment by methane** [comment]
I want API for nonblocking Read, or nonblocking Readable test.
Neither blocking wait nor callback don't satisfy my use case.

My use case is DB driver (go-sql-driver/mysql).
Basically, it is request-response protocol.  But sometime connection is closed by server or someone in the middle.
If we send query, we can't retry safely because query is not idempotent.  So we want to check connection
is closed or not before sending query.  Pseudo code is here.

```go
func (mc mysqlConn) Query(query string, args... driver.Value) error {
    if mc.conn.Readable() {
        // Connection is closed from server.
        return mc.handleServerGone()
    }

    mc.sendQuery(query, args...)
    return mc.recvResult()
}
```

Adding one or more goroutine makes the driver more complicated and slower.

**Comment by ianlancetaylor** [comment]
@methane I'm sorry, I don't understand your example: if `Readable` returns true, I would not expect the connection to be closed.  And, even if you omitted a `!`, if `Readable` returns false, that could just mean that the other side hasn't written any data; it doesn't mean that the connection is closed.  Determining whether it is possible to read from a socket is not the same as determining whether the socket is closed.

**Comment by methane** [comment]
@ianlancetaylor Note that how we get EOF.  We need to call `conn.Read()`.

As I said above, what I want is "nonblocking Read, or nonblocking Readable test."

**Comment by dvyukov** [comment]
Won't write on a closed/broken connection return an error? If yes, then you could just write the request always. If write succeeded, request was sent. If write failed, request wasn't send.
This has obvious races, but it's the same as for your code.

**Comment by methane** [comment]
@dvyukov No.  Even though remote peer closed connection, `conn.Write()` will success.
It's because TCP allows "half close".  (I hadn't tested it on Unix socket yet)

**Comment by methane** [comment]
@dvyukov DB closes full connection at socket API level.  But it doesn't affect TCP.
Until client closes connection (including half close of write side), `Write()` will success.

**Comment by methane** [comment]
@dvyukov Please try this code:
https://play.golang.org/p/ElUBU_l8QC2

**Comment by dvyukov** [comment]
Yup, write fails as expected:
```
$ go run /tmp/test.go
0 Read 0 (EOF)
0 Written 6 (<nil>)
1 Read 0 (EOF)
1 Written 0 (write tcp 127.0.0.1:42246->127.0.0.1:39591: write: broken pipe)
2 Read 0 (EOF)
2 Written 0 (write tcp 127.0.0.1:42246->127.0.0.1:39591: write: broken pipe)
```

I would expect it to be the other way around. Readability test should always check local read buffer first. So if some data is already received, but remote end is closed, readability test will always succeed, but write will fail.

**Comment by methane** [comment]
> Yup, write fails as expected:

```
0 Read 0 (EOF)
0 Written 6 (<nil>)
```

You send 6 bytes, even after you received EOF.

**Comment by dvyukov** [comment]
Why does kernel do this? Is it a kernel bug? The connection is already in CLOSE_WAIT state. This makes read fail, but not write...

**Comment by methane** [comment]
Because there are no difference between half close (`CloseWrite`) and full close on TCP layer.
Both send just one FIN packet.
So client don't know remote peer will receive data or not.

When client send data packet, server will return RST packet.
That's why second send returns EPIPE.  Here is tcpdump of the sample program.

```
tcpdump: listening on lo, link-type EN10MB (Ethernet), capture size 262144 bytes
19:45:09.516488 IP (tos 0x0, ttl 64, id 21231, offset 0, flags [DF], proto TCP (6), length 60)
    localhost.33704 > localhost.11451: Flags [SEW], cksum 0xfe30 (incorrect -> 0x751c), seq 30788114, win 43690, options [mss 65495,sackOK,TS val 4103978601 ecr 0,nop,wscale 7], length 0
19:45:09.516497 IP (tos 0x0, ttl 64, id 0, offset 0, flags [DF], proto TCP (6), length 60)
    localhost.11451 > localhost.33704: Flags [S.E], cksum 0xfe30 (incorrect -> 0x0a34), seq 2698581878, ack 30788115, win 43690, options [mss 65495,sackOK,TS val 4103978601 ecr 4103978601,nop,wscale 7], length 0
19:45:09.516505 IP (tos 0x0, ttl 64, id 21232, offset 0, flags [DF], proto TCP (6), length 52)
    localhost.33704 > localhost.11451: Flags [.], cksum 0xfe28 (incorrect -> 0xdcb8), ack 1, win 342, options [nop,nop,TS val 4103978601 ecr 4103978601], length 0
19:45:09.516529 IP (tos 0x0, ttl 64, id 36746, offset 0, flags [DF], proto TCP (6), length 52)
    localhost.11451 > localhost.33704: Flags [F.], cksum 0xfe28 (incorrect -> 0xdcb7), seq 1, ack 1, win 342, options [nop,nop,TS val 4103978601 ecr 4103978601], length 0
19:45:09.616632 IP (tos 0x2,ECT(0), ttl 64, id 21234, offset 0, flags [DF], proto TCP (6), length 58)
    localhost.33704 > localhost.11451: Flags [P.], cksum 0xfe2e (incorrect -> 0x9869), seq 1:7, ack 2, win 342, options [nop,nop,TS val 4103978701 ecr 4103978601], length 6
19:45:09.616641 IP (tos 0x2,ECT(0), ttl 64, id 0, offset 0, flags [DF], proto TCP (6), length 40)
    localhost.11451 > localhost.33704: Flags [R], cksum 0x4929 (correct), seq 2698581880, win 0, length 0
```

**Comment by dvyukov** [comment]
Interesting. Thanks.

But readability is also not what you want, right? Readability will say OK is connection is half closed the other way, or fully closed but have local read data queued.

**Comment by methane** [comment]
"Readability" meant "conn.Read() will return without block".
MySQL protocol is basically "request-response" protocol.  Until we send a query, server don't send
anything in most case.  But there are exceptions: server shut down, server (or Linux or router) closed idle connection, etc.

In such cases, we don't want send query because query may be not idempotent.
That's why I want non-blocking Read.

**Comment by DemiMarie** [comment]
If your query is not idempotent, you are sunk.  The connection might
disappear after the server has received the query and committed the result,
but before it can respond.  You just don’t know.

You can have at-least-once, or at-most-once.  Exactly-once has been proven
impossible.  It simply can’t be done, unless the recipient keeps a record
of which messages it has received, a sequence number, or something similar.

On Nov 30, 2018 8:01 AM, "INADA Naoki" <notifications@github.com> wrote:

"Readability" meant "conn.Read() will return without block".
MySQL protocol is basically "request-response" protocol. Until we send a
query, server don't send
anything in most case. But there are exceptions: server shut down, server
(or Linux or router) closed idle connection, etc.

In such cases, we don't want send query because query may be not idempotent.
That's why I want non-blocking Read.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/golang/go/issues/15735#issuecomment-443196520>, or mute
the thread

**Comment by methane** [comment]
@DemiMarie It's off topic.

**Comment by ianlancetaylor** [comment]
What if instead of an explicit callback API we use something channel based?  We could add a method `NotifyWhenReadable` to `os.File` and `net.TCPConn`, etc.

```Go
func (t *TCPConn) NotifyWhenReadable(c chan<- interface{}, val interface{})
```

When the connection is readable, `val` is sent to `c`.  This would happen once.  If you wanted another notification, you would have to call `NotifyWhenReadable` again.  Naturally the channel send would be non-blocking, and the channel should be buffered.

This is a little bit like `signal.Notify`.

This permits a callback API but doesn't guide people toward it.  A goroutine could wait for its connection to be readable easily enough via

```Go
    ch := make(chan interface{}, 1)
    conn.NotifyWhenReadable(ch, nil)
    <-ch
```

or you could have a single server goroutine

```Go
    for f := range ch {
        go f.(func())()
    }
```

and each separate connection would

```Go
    conn.NotifyOnRead(ch, func() { /* do the next thing */ })
    return
```

**Comment by dvyukov** [comment]
> Naturally the channel send would be non-blocking

How are lost values supposed to be handled? If we drop a notification, part of the program will stay deadlocked forever?

Here @prattmic says the reason gvisor avoids standard network polling is not memory consumption, but rather latency:
https://github.com/golang/go/issues/29734#issuecomment-457303059
So I think we need to consider this aspect as well and if possible not introduce inherent latency penalty.

**Comment by ianlancetaylor** [comment]
Since a notification is sent at most once per call to `NotifyWhenReadable`, it's easy to give the channel a large enough buffer to never lose any notifications.

I think gVisor is kind of a special case.  I think most large servers care more about memory.  But of course if we can fix both that would be better.

**Comment by dvyukov** [comment]
> Since a notification is sent at most once per call to `NotifyWhenReadable`, it's easy to give the channel a large enough buffer to never lose any notifications.

You mean for the channel per connection scheme? If a single channel is used for all connections, it would require knowing some hard upper bound on number of connections.

If one of the recommended schemes is to use a single channel, I suspect it will become a scalability bottleneck. Of course users can start partitioning the channel into N independent channels, but then it will have a problem with load balancing: either some partitions will be overloaded and/or bad locality and/or preallocataing N times more memory for channel buffers.

I also wonder if we will be able to do this "exactly one non-spurious notification". I think C networking servers will do own polling, but then always followed by non-blocking reads/writes. If we leave reads/writes blocking, then it will bring very unique requirements for the polling mechanism.

**Comment by dvyukov** [comment]
E.g. the current linux polling scheme is always armed edge-triggered epoll. It by design generates spurious notifications.

**Comment by gobwas** [comment]
Why don’t just add callback interface? That is, with the same semantics as @ianlancetaylor suggested – with re-register of the callback for the next readable event after one occurred. Inside callback user could define its own strategy of handling event. Probably it will dispatching task between some free goroutines. 

Also regarding this comment from @dvyukov:
> If we do OnReadable(f func()), won't we need to fork half of standard library for async style? Compress, io, tls, etc readers all assume blocking style and require a blocked goroutine.

I think this may not affect any wrappers around raw conn in general – we could subscribe for readable event on conn and read from wrapper, hoping that it would not block on split cases, when available data is not sufficient for wrapper purposes of decoding/processing data.

**Comment by ianlancetaylor** [comment]
@gobwas A callback interface is what @bradfitz suggested above.  My main concern with a callback interface is, simply, it's not typical Go style.  I was hoping to explore alternatives.

**Comment by bradfitz** [comment]
And to be clear to others (Ian knows this), in the same comment I suggested that above, I wrote:

> Yes, maybe this is getting dangerously into event-based programming land.

Which is another way of saying "not typical Go style"

**Comment by funny-falcon** [comment]
unfortunately, "not typical Go style" is the thing people have to fall in when they want top perofrmance and/or millions of connections per process.
https://medium.freecodecamp.org/million-websockets-and-go-cc58418460bb
https://github.com/eranyanay/1m-go-websockets

**Comment by funny-falcon** [comment]
typical Go style is quiteconvenient for high-level nd medium-levellogic, and quite inefficient at low level.

**Comment by gobwas** [comment]
@ianlancetaylor as exploration, maybe option of exporting network poller API would be better idea than a callback? I mean something like this:

```go
import "net/poll"

p, err := poll.Create(...) // epoll_create() on Linux.
if err != nil {
    // handle error
}
go func() {
    for ev := range poll.Wait() { // epoll_wait on Linux.
        ev.Conn.Read(...)
        ev.Data // User data bound within Notify() call.
    }
}()
conn, err := net.Dial(...)
if err != nil {
    // handle error
}
if err := p.Notify(poll.Read, conn, data); err != nil { // epoll_ctl on Linux.
    // handle error
}
```

Thus, we may leave general poller untouched (and non-blocking), but provide additional polling mechanism which will block when poll.Wait() stopped being drained.

Also Notify() api may differ, e.g. accepting different channels to write event to, as you suggested, but in blocking manner – without events loss and with strong notice in method comment =)

**Comment by networkimprov** [comment]
> I also wonder if we will be able to do this "exactly one non-spurious notification". I think C networking servers will do own polling, but then always followed by non-blocking reads/writes. If we leave reads/writes blocking, then it will bring very unique requirements for the polling mechanism.

@dvyukov, would it work to do a non-blocking read of a byte or two internally to verify readability before raising a notification? That adds at least one syscall to the overhead, but for a network or storage situation, maybe that's OK.

In the spirit of epoll, it would be helpful to get notifications for an arbitrary set of io.Reader objects via a single channel, since we cannot `select` on a runtime-defined set of channels.

**Comment by dvyukov** [comment]
> would it work to do a non-blocking read of a byte or two internally to verify readability before raising a notification?

This looks dirty and slow and in the end there can be just 1 byte. So it was readable, but not now anymore ;)

> In the spirit of epoll, it would be helpful to get notifications for an arbitrary set of io.Reader objects via a single channel, since we cannot `select` on a runtime-defined set of channels.

I think this is supported by the Ian's proposal, no?

**Comment by networkimprov** [comment]
Then let r.Read() be non-blocking after r.NotifyWhenReadable() and possibly return error io.WouldBlock?

**Comment by gobwas** [comment]
@networkimprov JFYI there is an ability to Select() on runtime defined set of channels via reflect api.

**Comment by networkimprov** [comment]
Do you have a plan to deal with the issue Dmitry raised?

**Comment by ianlancetaylor** [comment]
@networkimprov Which issue?  If you mean the problem with possibly losing a channel notification, then of course he is quite right, and the channel send must be a blocking send.  Depending on the implementation, this may require spinning up a goroutine to do the send.

**Comment by networkimprov** [comment]
No, I was referring to

> I also wonder if we will be able to do this "exactly one non-spurious notification". I think C networking servers will do own polling, but then always followed by non-blocking reads/writes. If we leave reads/writes blocking, then it will bring very unique requirements for the polling mechanism. E.g. the current linux polling scheme is always armed edge-triggered epoll. It by design generates spurious notifications.

How do you deal with this without letting r.Read() be non-blocking after r.NotifyWhenReadable() and possibly returning error io.WouldBlock?

**Comment by ianlancetaylor** [comment]
The problem I was trying to address was "large server with a lot of goroutines burns a lot of memory in buffers waiting to read."  For that purpose it's OK if a spurious notification leads to a blocking `Read`, as long as spurious notifications are rare.  But maybe I'm missing something.

**Comment by networkimprov** [comment]
Brad was trying to get rid of the idle goroutines: https://github.com/golang/go/issues/15735#issuecomment-266574151

If that's not a requirement, an alternate read API could __internally poll, malloc/free, and do non-blocking reads__, and return a new buffer on success. An internal buffer pool would avoid thrashing the allocator due to fruitless reads after spurious poll events.
```
func (r *T) ReadSize(maxLen int) (byte[], error)
```

**Comment by ianlancetaylor** [comment]
My channel suggestion also permits getting rid of the idle goroutines, for people who want to write a more complex program.

**Comment by networkimprov** [comment]
But it does incur a read-buffer + idle goroutine for every spurious poll event, which could be costly.

@dvyukov, any idea as to the proportion of spurious epoll wakeups in Linux?

More on epoll difficulties (from 2y ago):
https://news.ycombinator.com/item?id=13736674

**Comment by dvyukov** [comment]
> @dvyukov, any idea as to the proportion of spurious epoll wakeups in Linux?

No, I don't.
But if a solution relies on absence of spurious wakeups for correctness/performance, I guess we need these numbers for all OSes before making a decision.

**Comment by MaurGi** [comment]
So if I understand correctly, I am now stuck: I cannot read a 0 byte string to detect if the connection is closed and I cannot read a 1 byte because my server never writes anything and the read is blocking?

**Comment by methane** [comment]
FYI, we implemented this in mysql driver.
https://github.com/go-sql-driver/mysql/blob/master/conncheck.go

Note that it works only on Unix, and it caused several allocations.
I really want this is implemented in stdlib in efficient (and probably cross platform) way.

**Comment by detailyang** [comment]
It's highly useful to reduce overhead when the go program holds a lot of connections (connection pool) and now I'm using the `epoll` mechanisms do this which is not idiomatic.

Looking forward to land in Go, many thanks:)

**Comment by szuecs** [comment]
@ianlancetaylor a typical problem I have in an http proxy is, that connection spikes can create spikes in memory usage. I think this can be fixed with using epoll and I hope your approach will cover the problem. We would need to be able to set the max concurrency level for the goroutine calls, that will read and write from/to sockets.
We have an internal protection mechanism to avoid this problem, but you can see the memory spike in in bufio:

![image](https://user-images.githubusercontent.com/50872/60974223-01766100-a32a-11e9-83f0-c390be34af54.png)

**Comment by xtaci** [comment]
I need readability notification badly in my project, as I have to pre-allocate 4k buffer per connection.
https://github.com/xtaci/kcptun/blob/master/client/main.go#L56

**Comment by eloff** [comment]
So there is RawConn now which has an interface very much like what @bradfitz was proposing here. However, it calls the read callback _before_ calling wait for read. It must do this, as the net poller uses edge-triggered events - they won't fire if there is already data on the socket.

One workaround is to use a small stack buffer for the initial Read, and then when that reads some data, allocate and copy it to a real buffer, and then call Read again. That helps, but you'll still have the goroutine's 4KB stack overhead.

Another option is use RawConn.Control with `unix.SetsockoptInt(int(fd), unix.SOL_SOCKET, unix.TCP_DEFER_ACCEPT, 1)` for connections like HTTP where the client must send data first, it won't create a goroutine in the first place until there's data buffered, but only helps for that initial Read. It can be combined with the stack buffer approach for long-lived connections.

I'd still like to see an approach (maybe an async OnReadable callback method on RawConn, like bradfitz is proposing here?) to avoid having the goroutine overhead. Mail.ru avoids the net poller entirely and manually manages epoll for precisely this reason, as 4KB per WebSocket connection with millions of open but mostly idle WebSockets waiting for new mail is just too much overhead.

**Comment by xtaci** [comment]
> I'd still like to see an approach (maybe an async OnReadable callback method on RawConn, like bradfitz is proposing here?) to avoid having the goroutine overhead. Mail.ru avoids the net poller entirely and manually manages epoll for precisely this reason, as 4KB per WebSocket connection with millions of open but mostly idle WebSockets waiting for new mail is just too much overhead.

I don't think OnReadable callback is a good idea. If I have 2 handlers to toggle based on incoming data, then it's not impossible to code this kind of **nested callbacks** which reference to each other.

For such reason, in order to writing comprehensive logic we have to copy the buffer from callback out to another goroutine, in such case, the memory usage will be out of control, as we lost the ability to back-pressure the congestion signal to senders. (Before, we won't start `next net.Read()` until we've processed data.)

Even for a callback like `func() { Read);Process();Write() }` scenario, if Write() blocks, this callback still holds 4KB-buffer waiting to be sent.

In all these cases above, we still have inactive 4KB-buffers somewhere.

**Comment by xtaci** [comment]
I wrote a library based on ideas above, golang async-io

https://github.com/xtaci/gaio

**Comment by ortuman** [comment]
For my particular use case having a non-blocking `Read` is highly desirable. An chat service aiming to manage millions of connections per instance. Blocking interface forces to have an alive goroutine per connection, which makes this goal totally unrealistic.

**Comment by pmgexpo17** [comment]
> I wrote a library based on ideas above, golang async-io
> 
> https://github.com/xtaci/gaio

And it works great - it uses an epoll event handler and callback to solve the readability and writeability issues that are discussed here (if I'm not mistaken ... lol)
I use it instead of a multiplexer for a message proxy component which handles one to many client connections for inbound and outbound traffic.
It has a batch style of message processing which requires a message reader and frame size delimited message protocol. That is, the message reader compiles a complete message from 1 or more partial messages provided in each event result buffer

**Comment by lesismal** [comment]
I wrote another async-io lib [nbio](https://github.com/lesismal/nbio) to avoid using two goroutines per connection and reduce memory usage.
It supports http 1.x and is basically compatible with net/http. Many net/http-based web frameworks can be easily converted to use the async-io network layer.
It also supports tls and websocket.
I am trying to increase the load capacity of a single hardware, 1m/1000k-connections would not be that hard anymore for golang.

https://github.com/lesismal/nbio

**Comment by bcmills** [comment]
> In addition to the array memory backed by the slice given to Read, the goroutine itself is ~4KB of wasted memory.

If the problem is the size of the idle goroutine stack, would it make sense to have the runtime recognize goroutines that are likely to block ~immediately and allocate a much smaller initial stack for them?

**Comment by networkimprov** [comment]
@bcmills, maybe that could be a new proposal?

**Comment by funny-falcon** [comment]
> > In addition to the array memory backed by the slice given to Read, the goroutine itself is ~4KB of wasted memory.
> 
> If the problem is the size of the idle goroutine stack, would it make sense to have the runtime recognize goroutines that are likely to block ~immediately and allocate a much smaller initial stack for them?

Problems are both goroutine stack and buffer for Read. Especially when TLS is involved.

**Comment by ortuman** [comment]
At least, if we had goroutines memory pressure under control the read buffer issue may be tackled by extending `net.Conn` API:

```
conn := getTLSConn()

if err := conn.WaitForContent(); err != nil {
    return err
}
rdBuff := grabReadBuffer()
defer releaseReadBuffer(rdBuff)

n, err := conn.ReadAvailableContent(rdBuff)
if err != nil {
    return err
}
handleContent(rdBuff)
return nil
```

**Comment by lesismal** [comment]
> > Problems are both goroutine stack and buffer for Read. Especially when TLS is involved.
> 
> At least, if we had goroutines memory pressure under control the read buffer issue may be tackled by extending `net.Conn` API:
> 
> ```
> conn := getTLSConn()
> 
> _, err := conn.WaitRead() // block until there's available content
> if err != nil {
>     return err
> }
> rdBuff := grabReadBuffer()
> defer releaseReadBuffer(rdBuff)
> 
> n, err := conn.ReadAvailable(rdBuff)
> if err != nil {
>     return err
> }
> handleContent(rdBuff)
> return nil
> ```

I wrote an async-io lib [nbio](https://github.com/lesismal/nbio) and forked std tls and rewrote it to support non-blocking.
Here are examples that save more memory:
https://github.com/lesismal/nbio/pull/62#issuecomment-881221338

In my simple echo test with100k websocket tls connections, compared with std based frameworks:
| solution| qps | cpu | memory | 
|---|---|---|---|
| nbio | avg:110-120k, running fine | around 300% | 1.3G |
| std |  avg: 60-80k, with obvious stw | around 300% | 3.3G|

**Comment by CAFxX** [comment]
If I may voice a contrary opinion, I would argue that there are other ways to obtain the same benefits (minimize space wasted on parked goroutine stacks and on buffers that are waiting for data to be written into them) without unleashing callbacks and/or readability signals.

One such way could be to have buffers be decommitted from memory before submitting them for a blocking read (I haven't tested this, but AFAIK it should work). Right now this would require an additional syscall (e.g. madvise) before the read, but once/if #31908 comes around that won't be required anymore (as the madvise and read could be submitted in a single syscall). Given page-aligned buffers of size greater than a page, this should transparently achieve the goal at least for the buffers.

For the goroutine stack, something similarly transparent could also be achieved: as part of GC we already shrink the stacks of goroutines, if they are too big. It is not impossible to imagine to extend this mechanism to detect goroutines that have been parked for a while waiting for I/O and that would "freeze" their stacks, aggressively packing together the contents of the frozen stacks (without maintaining the minimum 4KB size), and freeing them. When the goroutine needs to be unparked a new stack of the appropriate size would be allocated, the frozen stack contents would be copied into the newly allocated stack and the goroutine would resume. This should transparently achieve the goal for the stacks of goroutines blocked on I/O.

I won't deny that there is a lot of handwaving in this comment: in my defense, I am not trying to provide a fully fleshed-out design. I'm just trying to point out that there may be other possibilities to achieve the same goals besides the mechanisms being discussed in the rest of this issue. Because these mechanisms, as has been pointed out already, are pretty non-go-like I would hope that any alternative that may achieve transparently the desirable goals set out in this issue is fully exhausted before more extreme options are implemented.

**Comment by dvyukov** [comment]
There are several potential issues with madvise:
1. It conflicts with large 2M pages.
2. The buffers are not necessary page-aligned.
3. The contents are not necessary discard-able (for a short read the rest of buffer must be left intact).

**Comment by CAFxX** [comment]
(sorry for being off topic, I will not add further comments in this issue; if needed I will update this comment to address further replies to avoid polluting the discussion with design considerations of a potential counterproposal)

IIUC 1 and 2 have multiple potential solutions (document that only pagesize-aligned buffers can benefit from this optimization, teach the memory allocator to handle pagesize-sized allocations especially, provide an allocator function specialized for this purpose, suggest to use off-go-heap memory, ...) so possibly these are not a showstopper (unless I'm missing something).

3 OTOH could be a showstopper (for the buffer part), but I'm confused. `io.Reader` docs seems to indicate that is not the case:

```
type Reader interface {
	Read(p []byte) (n int, err error)
}

[...] Even if Read returns n < len(p), it may use all of p as scratch space during the call. [...]
```

It only says that the space beyond n can be used as scratch space, but it does not say whether the contents are restored after the area has been used as scratch space. I always interpreted that to mean that data in p[n:] may not contain the original contents when `Read` returns, otherwise to be able to use all of `p` as scratch space `Read` would always require yet another scratch space to temporarily store a copy of `p` just in case of a short read (but then what would be the point of using p as the scratch space?). Did I get this wrong all this time?

**Comment by dvyukov** [comment]
> [...] Even if Read returns n < len(p), it may use all of p as scratch space during the call. [...]

I think you are right. I just assumed it should not change what's not returned following the principle of the least surprise.

**Comment by ivanjaros** [comment]
I am surprised to find out that there is no non-blocking way to read network connections in stdlib while Go is the most network-centric language out there. How come there is no progress since 2016 and one has to look for 3rd party libraries(none of which work on windows)?

**Comment by ianlancetaylor** [comment]
@ivanjaros The Go runtime uses non-blocking I/O internally, so that Go programs don't have to.

Note that this issue is not about non-blocking I/O.  It's about a way for a goroutine to wait for a connection to be readable without having to actually read from it.

**Comment by ivanjaros** [comment]
@ianlancetaylor my point was that this issue has been ongoing for years but we still have to result to using 3rd party libraries when we want to seek the bleeding edge of performance over DX through any of:

https://github.com/tidwall/evio
https://github.com/panjf2000/gnet
https://github.com/hslam/netpoll
https://github.com/xtaci/gaio
https://github.com/Allenxuxu/gev
https://github.com/pigogo/netgo
https://github.com/npat-efault/poller
https://github.com/alberliu/gn
https://github.com/smallnest/1m-go-tcp-server

On the other hand, since there are these options, the entire discussion might have become obsolete 🤷‍♂️

**Comment by lesismal** [comment]
@ivanjaros 
It seems only my lib [nbio](https://github.com/lesismal/nbio) supports async-io and async parser for tls/http/websocket by now but wasn't listed in your list and few people know it...

> Note that this issue is not about non-blocking I/O. It's about a way for a goroutine to wait for a connection to be readable without having to actually read from it.

@ianlancetaylor If go support a readable callback, the conn.Read and conn.Write should not blocked any more, else users still need to use at least one goroutine to handle Read/Write logic, then it would not be different from the current std solution on reducing the cost from lots of goroutines.

**Comment by lesismal** [comment]
> Sounds like you are getting close to #15021.
> 
> I'm worried that the existence of such a method will encourage people to start writing their code as callbacks rather than as straightforward goroutines.

@ianlancetaylor 


Definitely agree that async-io does need to be used callback in parsing logic, in which we also needs to consider the memory optimization of the half-packet bytes cache.
But only the part of parsing needs to use callbacks, such as the http parser.

The application-layer users can still keep the synchronous style as std does: async-io and callback-parser parse the complete package and pass them to a size-limited-goroutine-pool for business logic processing. 
Actually, my lib has already implemented it like this and basically compatible with net/http. A simple hello-world server as below:
```golang
package main

import (
	"fmt"
	"log"
	"net/http"

	"github.com/lesismal/nbio/nbhttp"
)

func HelloWorld(w http.ResponseWriter, r *http.Request) {
	fmt.Fprintf(w, "hello world")
}

func main() {
	router := &http.ServeMux{}
	router.HandleFunc("/", HelloWorld)
	// also can use gin
	// router := gin.New()
	// router.GET("/", ...)
	// or echo
	// router := echo.New()
	// router.GET("/hello", ...)

	svr := nbhttp.NewEngine(nbhttp.Config{
		Network: "tcp",
		Addrs:   []string{"localhost:8080"},
		Handler: router,
	})

	err := svr.Start()
	if err != nil {
		log.Fatalf("nbio.Start failed: %v\n", err)
		return
	}
	defer svr.Stop()

	<-make(chan int)
}
```

And even can be used with other std based frameworks like gin or echo, more examples:
- [gin-http-and-websocket-server](https://github.com/lesismal/nbio_examples/blob/master/http_with_other_frameworks/gin_server/gin_server.go)
- [echo-http-and-websocket-server](https://github.com/lesismal/nbio_examples/blob/master/http_with_other_frameworks/echo_server/echo_server.go)

I am sorry if it is not appropriate to mention my own lib here, but I think this [async-io] + [callback parsing] + [goroutine pool] solution can really allow us both to get the resource cost under control and to write synchronous logic.
We have the cake and eat it, too.
It has been verified in load testing and in some of my projects.

**Comment by ianlancetaylor** [comment]
There is waiting for a TCP connection to have data available, which is this issue.  And there is non-blocking I/O.  They are not the same thing.

**Comment by lesismal** [comment]
> There is waiting for a TCP connection to have data available, which is this issue. And there is non-blocking I/O. They are not the same thing.

We know that it is blocking io, but the currently provided interfaces such as conn.Read/Write are non-blocking interfaces for the application layer.
If the non-blocking Read/Write interface is not provided with Readable together, for the application layer, it is the same as the current std.

**Comment by szuecs** [comment]
@lesismal can you provide the data, maybe it's more convincing to show data. 

@ianlancetaylor  If we would consider to have a fixed size goroutine pool (as suggested in https://github.com/golang/go/issues/15735#issuecomment-1235331181) for handlers that can be set by some configuration, it would also fix the unbounded memory usage issue in https://github.com/golang/go/issues/35407.

**Comment by lesismal** [comment]
@szuecs I've provided some test results on the previous floor, please check it or better to try the code in yourself's env:
https://github.com/golang/go/issues/15735#issuecomment-917435376

**Comment by ianlancetaylor** [comment]
> We know that it is non-blocking io, but the currently provided interfaces such as conn.Read/Write are blocking interfaces for the application layer.

As you probably know, they block the goroutine that calls them.  They don't block any threads.  Goroutines are cheap.  Programs that need to avoid blocking goroutines are a special case.  And as noted above, it is already possible to use non-blocking I/O in Go.

**Comment by lesismal** [comment]
> As you probably know, they block the goroutine that calls them. They don't block any threads. Goroutines are cheap. Programs that need to avoid blocking goroutines are a special case. And as noted above, it is already possible to use non-blocking I/O in Go.

Yes, I know that they don't block any threads.
But when using std and there are millions of connections, the huge num of goroutines are not as `cheap` as there are only thousands of connections. That's why there are many third libs to do it as @ivanjaros listed:
https://github.com/golang/go/issues/15735#issuecomment-1235148890

**Comment by lesismal** [comment]
In a 1m connections test:
Golang needs at least 8G memory(8k*1m), and the CPU cost for both schedule and GC is also huge, so we can't deploy this service on a VM/hardware with low-middle level specifications, else we come across both OOM and STW problems.
Compared to other languages in this test, Golang performs even worse than Java-Netty and Nodejs, with multiple memory costs, that's such unexpected `amazing`.

As a cloud-native language, Golang will handle more and more connections as basic frameworks. The more we can reduce the num of goroutines, the more hardware, energy, and money we can save. 

Not joking, but sincerely:joy::
For environmental protection!
For the earth!

**Comment by lesismal** [comment]
My opinion is that if provide `Readable`, non-blocking Read/Write interface should be provided together, else `Readable` will not be useful.

**Comment by ivanjaros** [comment]
When PHP 7, i think, came out couple of years ago, Rasmus(PHP author) had a talk on some PHP conference where he showed how much electricity the better performance of this new version saved due to lower demand on hardware and hence lower economic costs in general. So this is not as outlandish as it might seem. As pointed above, the cost-savings are already impressive https://github.com/golang/go/issues/15735#issuecomment-917435376 and with stdlib implementation we might get even better CPU performance.

**Comment by d2gr** [comment]
All of the implementations listed above use mostly a reactor pattern, not a pro-actor pattern. See [proactor pattern](https://en.wikipedia.org/wiki/Proactor_pattern). (boost::asio is a pro-actor lib).

If I am not mistaken the proposal here is to add readability which would be implemented using callbacks, which would make Go support some kind of semi-proactor interfaces. In my opinion this is fine, many people have been looking forward to this implementation, but I think the main problem is going to be the callbacks.
In a high-frequency environment where we call `OnReadable` more than 1M times per second, we are going to create a function object every time (because closures are structs), which will be allocated in the heap because it'll escape the stack.

I have implemented myself a library that uses a true pro-actor pattern (not open-sourced yet), and it performs well, but not better than Go. Mainly because the heap allocations increase and decrease per second (given the above callback problem).
Libraries that implement a pro-actor pattern might improve the TLS & HTTP/2 performance given that the `tls` package has locks everywhere and HTTP/2 in Go is full of channels and mutexes (not criticism, both libraries are very well done).

Summary: If this issue goes forward and a PR is presented, are we going to see closures allocated in the stack? In C++, lambdas are just objects (like in Go) but they are moved from stack to stack, afaik they are never allocated in the heap (unlike Go).
I know that callbacks might not allocate actual heap, they just reuse memory because if the object is less than 32KiB mallocgc takes memory that is available, but still reducing performance.

**Comment by ivanjaros** [comment]
I don't think we need callbacks. I think the following would do:

```
type HServer struct {
	addr string
	conns map[net.Conn]struct{}
	done context.Context
	addQueue chan net.Conn
	closeQueue chan net.Conn
}

func (s *HServer) Run() {
	lis, _ := net.Listen("tcp", s.addr)

       go s.loop()

	for {
		if conn, err := lis.Accept(); err != nil {
				s.addQueue <- conn
		}
	}
}

func (s *HServer) loop() {
	for {
		select {
			case <-s.ctx.Done():
				for conn := range s.conns {
					conn.Close()
				}
				s.conns = nil
				return
			case conn := <-s.addQueue:
				s.conns[conn] = struct{}{}
			case conn := <-s.closeQueue:
					conn.Close()
					delete(s.conn, conn)
			default:
				for conn := range s.conns {
					if conn.CanRead() { <----- this is what it's all about, right?
						go s.handle(conn)
					}
				}
		}
	}
}

func (s *HServer) handle(conn net.Conn) {
	buf := bytes.NewBuffer(nil)
	if _, err := buf.ReadFrom(conn); err != nil {
		println(err)
		s.closeQueue <- conn
		return
	}
	// ... do something with the data
}
```

**Comment by d2gr** [comment]
> I don't think we need callbacks. I think the following would do:
>
> ...

Why would you do that in an example? You don't want to condition your `handle`. In your code:
```
for conn := range s.conns {
	if conn.CanRead() { <----- this is what it's all about, right?
		go s.handle(conn)
	}
}
```
The conn.CanRead will condition the handling, so, you don't handle the conn? Ah no, you handle it but after some loops... Wasting CPU polling the channels and then falling through the default.

But that's why Go is slow. If you use one channel is fine, if you are polling multiple channels and expecting the software to be latency sensitive you are not going to achieve your goals. That's why libraries like fasthttp or evio were created. fasthttp uses standard Go but replaces the http package, and evio is just a replacement of net (with a reactor pattern). The problem that people want to solve is handling as many connections as possible with the lowest latency possible. And in standard Go that's not possible, mainly because you have a coroutine per client (maybe 2, one for reading and another for writing).

 The idea is to have a reactor patter, so:
```
func broadcast(msg []byte, conns []net.Conn) {
  for _, c := range conns {
    c.OnWritable(func() {
      c.Write(msg)
    })
  }
}
```
But the problem with the above is that `c.Write` might block. Because let's say msg is 65536 bytes, and the OS buffer is only able to hold 1024 bytes now. We'll write 1024 bytes and then... what? Some Go libraries might fail with an ErrShortWrite error (see [bufio](https://cs.opensource.google/go/go/+/refs/tags/go1.19.1:src/bufio/bufio.go;l=631;drc=54182ff54a687272dd7632c3a963e036ce03cb7c)). Therefore the `c.Write(msg)` must block. UNLESS you can specify to not block with something like `c.WriteNonblock(msg)`, but that would require A LOT of changes.

Given the above, and my previous comment, I think implementing this in std Go would be difficult. Not difficult, but it would require a lot of changes to the interfaces and existing data structures. It'd be nice to have a reactor patter library supported by Go std that uses Go data structures because they can make use of compiler directives to not be stored in the heap (See [this](https://github.com/golang/go/blob/master/src/runtime/HACKING.md#unmanaged-memory)). So suggesting a package like `net/reactor` wouldn't be that crazy, I think. I'd happily work on it

**Comment by ivanjaros** [comment]
You can simply but **continue** into the **default** statement to check the channels again once you handle all connections. But that was just an example. As for handling - the main point is to use goroutine only if there is data on the connection to be handled and not to have a dedicated goroutine per each connection. so if you have 100k connections but only 100 are active at any one moment, that is savings of 99 900 goroutines * 2kB(minimum) = 199 800 kB or roughly 200 MB memory they **would** take + context switching.

**Comment by lesismal** [comment]
@ivanjaros 

If `conn.CanRead()` is blocking interface, one connection's blocking makes all the other connections wait for long in the loop, even if they have been readable already.
If `conn.CanRead()` is non-blocking interface, that for loop will cause cpu 100%.

**Comment by ivanjaros** [comment]
true

**Comment by ivanjaros** [comment]
..although you could have have a connection collection primitive that would handle this internally and would work in a channel-like manner. so handing a single connection could work as blocking Read() and non-blocking CanRead(), neither would break existing api, and this new collection would be a new primitive extending the net package and which would allow to add and remove connection and block until one of the connections is readable.

**Comment by d2gr** [comment]
> ..although you could have have a connection collection primitive that would handle this internally and would work in a channel-like manner. so handling a single connection could work as blocking Read() and non-blocking CanRead(), neither would break existing api, and this new collection would be a new primitive extending the net package and which would allow to add and remove connection and block, with timeout, until one of the connections is readable, in which case it would return it for reading. ...just thinking out loud here.

Too complex. You don't want some things explicitly blocking and others implicitly, although the doc makes a distinction. IMO what people want is 2 different approaches, one more pro-actor that allows you to have one goroutine to handle multiple connections, and another (that already exists) to handle a connection per goroutine (go net). But the problems presented above might be a blocker.

**Comment by ivanjaros** [comment]
Either way, i just wanted to keep the conversation going since I'd like to write a chat server in near future and this type of functionality will be a must so I will have to use one of the libraries I listed above, but would prefer this to be in the stdlib.

Also I bet proxy like Traefik would get a nice performance gain from this as well. I mean, gnet equals in performance the fastest code out there in c/cpp/rust/...  so the performance is there, in Go. It just needs to be unlocked.

**Comment by lesismal** [comment]
>  I mean, gnet equals in performance the fastest code out there in c/cpp/rust/... so the performance is there, in Go

@ivanjaros 
That's not right. 
For HTTP, gnet uses a simple parser that does not implement full features of the HTTP protocol, so its testing code cost much less of cpu than a full-featured HTTP server.

As we know, golang can't get the same performance as c/cpp/rust but goroutine and chan make us to write code easier.

**Comment by d2gr** [comment]
> Also I bet proxy like Traefik would get a nice performance gain from this as well. I mean, gnet equals in performance the fastest code out there in c/cpp/rust/... so the performance is there, in Go. It just needs to be unlocked.

Yeah, but the problem with frameworks like gnet or evio is that they are not usable in production. There might be cases where you can use it, even with `codecs`. They are too complicated to be used in serious environments. A pro-actor pattern library would be more suitable (like boost::asio).

Standard Go is quite performant, just take a look at fasthttp. It is built using standard Go. The only problem with Go is the amount of sync mechanisms that it requires. That's why I said that a pro-actor pattern might be faster for TLS & HTTP/2. Or any protocol that uses streams instead of one single stateless connection.

**Comment by d2gr** [comment]
> As we know, golang can't get the same performance as c/cpp/rust in most scenarios, but can get near to c/cpp/rust in some scenarios such as IO, and the most important is: goroutine and chan make us write code easier.

I mean... Go is quite fast. I benchmarked fasthttp vs boost::beast quite a lot (in AWS 2xc5n.4xlarge, client and server in a placement group) and Go is able to handle 200K QPS below 5ms (100th percentile) and boost::beast was able to do 200K sometimes below 4.8ms other times below 5.5ms. So there's a lot of variance in boost::beast, whilst in Go I got the same result in all the benchmarks.
Now, the only problem with Go is that the more the connections the less performant the I/O becomes. And that's a limitation you should be aware of while building a system.

So it depends on how your Go program is structured. If it has locks, it doesn't. If uses channels, etc... In my benchmarks I had no locks, just plain HTTP (no TLS) with some caching system in both fasthttp & boost::beast.

I wouldn't say that Go is less performant than C++ and Rust. It also depends on the library that you use. boost::beast is ok, it does scale well, but it is not as latency sensitive as you might think (truth is that you can easily plug in solarflare's onload in boost::asio and get a performance improvement), and Rust is not as different from Go. They also use coroutines and they also need to lock in some scenarios but they have the advantage of not having a GC ("advantage").

Bechmarks are mostly a lie. People prepare their programs for the benchmark in question (like gnet's case). Production ready environments don't need to lie in their benchmarks ([fasthttp](https://github.com/TechEmpower/FrameworkBenchmarks/blob/master/frameworks/Go/fasthttp/src/main.go))

**Comment by lesismal** [comment]
@d2gr 
I think proactor is not the key point, but are the num of goroutines, and blocking or non-blocking.
I've tried a lot about a non-blocking based HTTP server, I think we can't gain both high performance and high online together using golang by now:
1. For std-based frameworks that use net.Conn(blocking IO interface), including fasthttp, the cost of hardware grows fast as the num of connections increases, because they all use at least one goroutine for each connection, including fasthttp. It's hard to reduce the cost for gc, memory and schedule.
2. For non-blocking frameworks, no matter whether it's reactor or proactor, we need to handle IO and logic in different goroutine pools. That needs more heap escape, more complex async parser, and we can not optimize the buffers as fasthttp does.

**Comment by d2gr** [comment]
@lesismal 
I mention pro-actor because it's the easiest way to handle async (unlike async/await, or reactor). I agree that we need less heap and more stack-based structures. I don't know what do you mean with `async parser`.

**Comment by ivanjaros** [comment]
> @ivanjaros That's not right. For HTTP, gnet uses a simple parser that does not implement full features of the HTTP protocol, so its testing code cost much less of cpu than a full-featured HTTP server.

you are mixing apples and oranges here. nobody is talking about http servers here. gnet is merely networking framework(like ALL the projects mentioned before). what you build on top of it is up to you.

**Comment by lesismal** [comment]
> 

@d2gr 
We need to cache half-packet bytes because we can't use `ReadFull`. The parser and buffer usage logic are much more complex than net.Conn based frameworks.

@ivanjaros Please refer to the reasons I've mentioned here and in previous comments.

Here are some benchmark reports, you can run the test in your own env:
https://github.com/lesismal/go-net-benchmark/issues/1
I get the same level of performance between nbio and gnet, but I support a lot more in nbio than gnet. I tried a lot to optimize the performance and reduce the cost, but I can achieve only balance between performance and cost reduce, can't gain both of them together. 
That's not only about  HTTP!
For simple IO logic, we do gain a good performance, but for product env, for the complex logic, the reasons I mentioned make the performance down. 
You can try it, I will be glad to see if you can find some way to make more promotions.

**Comment by bcmills** [comment]
@lesismal, @ivanjaros, @d2gr: it isn't clear to me how the above discussion relates to the feature proposed in this issue. For off-topic performance discussions, please start a thread on the `golang-dev` mailing list or a similar venue outside of the issue tracker.

**Comment by d2gr** [comment]
@bcmills Sorry for spamming a bit, but my comments where related to the issue. At least this one https://github.com/golang/go/issues/15735#issuecomment-1244473297

**Comment by lesismal** [comment]
@bcmills 
Sorry about that. 
But I think the `CanRead` does related to non-blocking interfaces and performance. If the new feature doesn't consider these points, the new feature will not be useful and should not be added.
If the new feature provides only a `CanRead` but still blocking `Read/Write`, there will be problems like @ivanjaros 's for-loop or gobwas/ws.

My previous related comments:

> My opinion is that if provide Readable, non-blocking Read/Write interface should be provided together, else Readable will not be useful.

> If `conn.CanRead()` is blocking interface, one connection's blocking makes all the other connections wait for long in the loop, even if they have been readable already. If `conn.CanRead()` is non-blocking interface, that for loop will cause cpu 100%.

**Comment by lesismal** [comment]
@bcmills 
Actually, `CanRead` interface is just the same thing that [1m-go-websockets](https://github.com/eranyanay/1m-go-websockets) and [gobwas/ws](https://github.com/gobwas/ws) did. That's the smallest import for event-driven on std's TCPConn, but as we've discussed, it leads to the problem:
https://github.com/gobwas/ws/issues/143
That's the same paroblem as the `for-loop-block` we discussed in previous comments.
To solve this problem in `gobwas/ws`, we still need to serve each connection with at least 1 goroutine, which come back to the solution of the current std but is more complex. Then, there's no benefit, seems and performs even worse.
So, if `Read/Write` interfaces are still blocking mod, I would prefer you maintainers just keep it not changed rather than add a new `CanRead` interface.

Also that's why I hope if you add `CanRead`, please add non-blocking `Read/Write` interfaces together. And then, one more thing, the non-blocking interfaces and seperate goroutine pool make the performance worse than current std if there are not a lot of connections, I think that's should be also considered before this proposal is accepted.
But all right, I will stop discussing performance, I just focus on the `CanRead`, blocking or non-blocking.

**Comment by ivanjaros** [comment]
what if the only addition into the stdlib for net.Conn would be read method that would behave like a channel? that way you can decide to wait for data or move elsewhere. so to use my previous example:

```
type HServer struct {
	addr string
	conns map[net.Conn]struct{}
	done context.Context
	addQueue chan net.Conn
	closeQueue chan net.Conn
}

func (s *HServer) Run() {
	lis, _ := net.Listen("tcp", s.addr)

       go s.loop()

	for {
		if conn, err := lis.Accept(); err != nil {
				s.addQueue <- conn
		}
	}
}

func (s *HServer) loop() {
	for {
		select {
			case <-s.ctx.Done():
				for conn := range s.conns {
					conn.Close()
				}
				s.conns = nil
				return
			case conn := <-s.addQueue:
				s.conns[conn] = struct{}{}
			case conn := <-s.closeQueue:
					conn.Close()
					delete(s.conns, conn)
			default:
				for conn := range s.conns {
                                        select {
                                             case data, err := <-conn.ChannelRead():
                                                delete(s.conn, conn) // remove from list
						go s.handle(conn, data) // handle connection with first chunk of data
                                             default:
                                             // continue to next connection
                                        }
				}
		}
	}
}

func (s *HServer) handle(conn net.Conn, data []byte) {
	buf := bytes.NewBuffer(data)
	if _, err := buf.ReadFrom(conn); err != nil {
		println(err)
		s.closeQueue <- conn
		return
	}
	// ... do something with the data
        conn.Write() // send some response
        // return connection back to "queue"
       s.addQueue <- conn
}
```

**Comment by AnimusPEXUS** [comment]
just add those functions to net.Conn:
```
ReadAvailable() bool
WriteAvailable() bool
ReadNonBlocking(b []byte) (n int, err error)
WriteNonBlocking(b []byte) (n int, err error)
```

**Comment by ianlancetaylor** [comment]
@AnimusPEXUS Thanks, but that doesn't address this issue.  This issue is about putting a goroutine to sleep until there is something to read.

**Comment by AnimusPEXUS** [comment]
@ianlancetaylor determining "if ther's something to read" is system-to-system specific question. for NIX* it's 'select' function. maximum you can do here for nix is make callback or push signal after select returned and if we want select non-blocking, then we have to put socket into non-blocking state. for non-socket/non-unix there only two options: endless loop, constantly checking readiness, or, again, callback if peer can somehow separately indicate 'I have something to read'

**Comment by DemiMarie** [comment]
I don’t think Go as it stands today is the best choice for this kind of code.

Optimized Rust and C++ servers use a combination of compiler-generated state machines (`async`/`await` and `co_async`/`co_await`, respectively) and manual memory management to achieve very high performance.  Go does not support either of these features, so achieving similar performance in Go would require essentially writing C with Go syntax.

**Comment by ianlancetaylor** [comment]
@AnimusPEXUS There is a reason that this issue is still open.

**Comment by black23eep** [comment]
Is there any progress on this issue now?

**Comment by scratchmex** [comment]
why can't we have a chan-like interface to use in select? like proposed in https://github.com/golang/go/issues/15735#issuecomment-286776017. This to have a single loop for send and receive on a connection

```go
// send and recv loop
for {
	select {
	case p <- SendChan:
		conn.Write(...)
	case <-conn.Readable():
		conn.Read(...)
	}
}
```

**Comment by ianlancetaylor** [comment]
It's difficult to implement efficiently in the runtime. At least, I don't see how to do it.

Note that you can already do something similar by having a goroutine read data from the network connection and send it on a channel. Yes, that requires an extra goroutine.

**Comment by gobwas** [comment]
> Yes, that requires an extra goroutine.

And an extra memory buffer – which I think was the reasoning for this feature request.

upd. My comment above is just to highlight that mentioned channel and a goroutine did basically a full circle in this discussion :)

**Comment by DemiMarie** [comment]
The options I know of are (from fastest to slowest):

1. Use DPDK/XDP and implement a userspace TCP stack.
2. Use a vendor-specific stack that uses kernel bypass to emulate 3 or 4.
3. Use io_uring with preregistered buffers.
4. Use the sockets API.

For maximum performance while still not relying on elevated privileges, I think it would be best to use a proactor model with `io_uring`.  If CGo were faster, binding to an optimized C or Rust library for this would be the obvious choice, at least if there were not lifetime constraints that forced the Go bindings to be highly unsafe.
